File: src/chunking/ParserPool.ts
/**
 * 解析器池管理
 *
 * 按语言缓存 Parser 实例，避免重复初始化。
 * 支持动态加载语言语法包。
 *
 */
import Parser from '@keqingmoe/tree-sitter';

// 语言到语法模块的映射
const GRAMMAR_MODULES: Record<string, string> = {
  typescript: 'tree-sitter-typescript',
  javascript: 'tree-sitter-javascript',
  python: 'tree-sitter-python',
  go: 'tree-sitter-go',
  rust: 'tree-sitter-rust',
  java: 'tree-sitter-java',
  c: 'tree-sitter-c',
  cpp: 'tree-sitter-cpp',
  c_sharp: 'tree-sitter-c-sharp',
};

// 缓存已加载的语法
// tree-sitter Language 类型是原生对象，没有导出类型定义
type TreeSitterLanguage = unknown;
const loadedGrammars: Map<string, TreeSitterLanguage> = new Map();

// 缓存已初始化的解析器
const parserCache: Map<string, Parser> = new Map();

/**
 * 加载指定语言的语法
 *
 * tree-sitter 0.20.x 的导出格式：
 * - tree-sitter-typescript: { typescript, tsx } (需要取 .typescript)
 * - 其他语言包: default export 直接是 Language 对象
 */
async function loadGrammar(language: string): Promise<TreeSitterLanguage | null> {
  // 检查缓存
  const cached = loadedGrammars.get(language);
  if (cached) return cached;

  const moduleName = GRAMMAR_MODULES[language];
  if (!moduleName) return null;

  try {
    // 动态导入语法模块
    const grammarModule = await import(moduleName);

    let grammar: TreeSitterLanguage | null = null;

    // tree-sitter-typescript 包特殊处理（包含 typescript 和 tsx 两个语言）
    if (language === 'typescript') {
      grammar = grammarModule.default?.typescript ?? grammarModule.typescript;
    } else {
      // 其他语言包: 0.20.x 版本直接使用 default export
      const exported = grammarModule.default ?? grammarModule;

      // 判断是否是有效的 Language 对象（有 nodeTypeInfo 属性）
      if (exported && typeof exported === 'object' && 'nodeTypeInfo' in exported) {
        grammar = exported;
      }
      // 尝试 language property（某些版本使用这种格式）
      else if (exported?.language) {
        grammar = exported.language;
      }
      // 尝试以语言名命名的属性
      else if (exported?.[language]) {
        grammar = exported[language];
      }
    }

    if (!grammar) {
      console.error(
        `[ParserPool] Could not extract grammar for ${language} from module ${moduleName}`,
      );
      return null;
    }

    loadedGrammars.set(language, grammar);
    return grammar;
  } catch (err) {
    console.error(`[ParserPool] Failed to load grammar for ${language}:`, err);
    return null;
  }
}

/**
 * 获取指定语言的解析器
 * @param language 语言标识
 * @returns Parser 实例，如果不支持该语言则返回 null
 */
export async function getParser(language: string): Promise<Parser | null> {
  // 检查缓存
  const cached = parserCache.get(language);
  if (cached) return cached;

  // 加载语法
  const grammar = await loadGrammar(language);
  if (!grammar) return null;

  // 创建解析器
  // grammar 来自 tree-sitter 原生模块（无导出类型，见 loadGrammar），断言到 setLanguage 期望的入参类型
  const parser = new Parser();
  parser.setLanguage(grammar as Parameters<Parser['setLanguage']>[0]);

  // 缓存
  parserCache.set(language, parser);
  return parser;
}

/**
 * 检查是否支持指定语言
 */
export function isLanguageSupported(language: string): boolean {
  return language in GRAMMAR_MODULES;
}

File: crates/chunker/src/parser_pool.rs
//! Port of `src/chunking/ParserPool.ts`.
//!
//! Maps a language id to its tree-sitter grammar and parses source. tree-sitter
//! reports byte offsets, so downstream `SourceAdapter` resolves the utf8 domain.
//!
//! Supported: typescript / javascript / python / rust / go / java / c / cpp /
//! c_sharp / ruby / php / shell(bash). The TS `GRAMMAR_MODULES` also lists
//! kotlin / swift / lua; their Rust grammar crates have non-standard versions
//! and are deferred (fall through to plain-text split).

use tree_sitter::{Language, Parser, Tree};

fn language_for_impl(language: &str) -> Option<Language> {
    match language {
        "typescript" => Some(tree_sitter_typescript::LANGUAGE_TYPESCRIPT.into()),
        "javascript" => Some(tree_sitter_javascript::LANGUAGE.into()),
        "python" => Some(tree_sitter_python::LANGUAGE.into()),
        "rust" => Some(tree_sitter_rust::LANGUAGE.into()),
        "go" => Some(tree_sitter_go::LANGUAGE.into()),
        "java" => Some(tree_sitter_java::LANGUAGE.into()),
        "c" => Some(tree_sitter_c::LANGUAGE.into()),
        "cpp" => Some(tree_sitter_cpp::LANGUAGE.into()),
        "c_sharp" => Some(tree_sitter_c_sharp::LANGUAGE.into()),
        "ruby" => Some(tree_sitter_ruby::LANGUAGE.into()),
        "php" => Some(tree_sitter_php::LANGUAGE_PHP.into()),
        "shell" => Some(tree_sitter_bash::LANGUAGE.into()),
        _ => None,
    }
}

/// Exposed to the symbols module for building tags.scm queries.
pub(crate) fn language_for(language: &str) -> Option<Language> {
    language_for_impl(language)
}

// Used at P3 integration (processor.ts decides AST vs fallback path).
#[allow(dead_code)]
pub fn is_language_supported(language: &str) -> bool {
    language_for(language).is_some()
}

/// Parse `code` for `language`. Returns None if the grammar is unknown or
/// parsing fails (caller falls back to plain-text splitting).
pub fn parse(language: &str, code: &str) -> Option<Tree> {
    let lang = language_for(language)?;
    let mut parser = Parser::new();
    parser.set_language(&lang).ok()?;
    parser.parse(code, None)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_rust() {
        let tree = parse("rust", "fn add(a: i32, b: i32) -> i32 { a + b }").unwrap();
        assert_eq!(tree.root_node().kind(), "source_file");
    }

    #[test]
    fn parses_python() {
        let tree = parse("python", "def f(x):\n    return x\n").unwrap();
        assert_eq!(tree.root_node().kind(), "module");
    }

    #[test]
    fn parses_typescript() {
        let tree = parse("typescript", "function add(a: number) { return a; }").unwrap();
        assert_eq!(tree.root_node().kind(), "program");
    }

    #[test]
    fn unknown_language_returns_none() {
        assert!(parse("brainfuck", "+++").is_none());
        assert!(!is_language_supported("brainfuck"));
    }

    #[test]
    fn parses_newly_added_grammars() {
        let cases: &[(&str, &str, &str)] = &[
            ("javascript", "function f(){return 1;}", "program"),
            ("go", "package main\nfunc main() {}\n", "source_file"),
            ("java", "class A { void m() {} }", "program"),
            ("c", "int main() { return 0; }", "translation_unit"),
            ("cpp", "int main() { return 0; }", "translation_unit"),
            ("c_sharp", "class A { void M() {} }", "compilation_unit"),
            ("ruby", "def f\n  1\nend\n", "program"),
            ("php", "<?php function f() { return 1; }", "program"),
            ("shell", "f() { echo hi; }\n", "program"),
        ];
        for (lang, code, root_kind) in cases {
            let tree = parse(lang, code).unwrap_or_else(|| panic!("parse failed for {lang}"));
            assert_eq!(
                tree.root_node().kind(),
                *root_kind,
                "root kind mismatch for {lang}"
            );
            assert!(is_language_supported(lang));
        }
    }
}
